### ISSUE TITLE
Fetch mechanics broken in develop for rare but present cases

### ISSUE THREAD

**Issue by chissg** [spack/spack#8733]
When attempting to stage or install a package depending on a package where at least one installed version can be found via packages.yaml _but which does not have a default URL or an explicit URL for the version_, an error stack like the following will result instead of a successful build (regardless of whether the errant version satisfies the top-level package's requirements or not):
```
==> Error: Class constructor failed for package 'sqlite'.

Caused by:
TypeError: expected string or buffer
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/repo.py", line 833, in get
    return package_class(spec)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/package.py", line 672, in __init__
    f = fs.for_package_version(self, self.version)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/fetch_strategy.py", line 977, in for_package_version
    url = pkg.url_for_version(version)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/package.py", line 794, in url_for_version
    default_url, self.url_version(version))
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/url.py", line 845, in substitute_version
    ver,  vs, vl, voffs) = substitution_offsets(path)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/url.py", line 789, in substitution_offsets
    ver,  vs, vl, vi, vregex = parse_version_offset(path)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/url.py", line 408, in parse_version_offset
    path, ext, suffix = split_url_extension(path)
  File "/home/greenc/work/cet-is/sources/spack-test/lib/spack/spack/url.py", line 346, in split_url_extension
    match = re.search(r'(.*(?:sourceforge\.net|sf\.net)/.*)(/download)$', path)
  File "/usr/lib64/python2.7/re.py", line 142, in search
    return _compile(pattern, flags).search(string)
```

It appears that Spack is going through the fetch strategy mechanics even though (a) they're not necessary because this is a version that's already installed, and (b), the lack of either a specific or default URL for this version means that it can never be fetched anyway.

### Steps to reproduce the issue

First, ensure that your system has an installed SQLite and that it is mentioned in `packages.yaml`, like (for instance for Scientific Linux 7.4):
```yaml
packages:
  sqlite:
    paths:
      sqlite@3.7.17 arch=linux-rhel7-x86_64: /usr
```

Then, attempt to install a package requiring SQLite, _e.g._:
```console
$ spack install openslide
```

### Information on your system

Scientific Linux 7.4, Spack from `upstream/develop`,  `packages.yaml` as described above.

**Comment by balay** [comment]
I'm seeing a similar problem  - so adding it to this issue tracker

I have in ~/.spack/packages.yaml

```
  intel-mkl:
    paths:
      intel-mkl@18.0.2.199: /opt/intel
    buildable: False
```
Now I get:
```
balay@cori03:~/spack-tmp> spack install --dirty hypre
==> Error: Class constructor failed for package 'intel-mkl'.

Caused by:
TypeError: expected string or buffer
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/repo.py", line 833, in get
    return package_class(spec)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/package.py", line 504, in __init__
    f = fs.for_package_version(self, self.version)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/fetch_strategy.py", line 976, in for_package_version
    url = pkg.url_for_version(version)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/package.py", line 626, in url_for_version
    default_url, self.url_version(version))
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/url.py", line 845, in substitute_version
    ver,  vs, vl, voffs) = substitution_offsets(path)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/url.py", line 789, in substitution_offsets
    ver,  vs, vl, vi, vregex = parse_version_offset(path)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/url.py", line 408, in parse_version_offset
    path, ext, suffix = split_url_extension(path)
  File "/global/u2/b/balay/spack-tmp/lib/spack/spack/url.py", line 346, in split_url_extension
    match = re.search(r'(.*(?:sourceforge\.net|sf\.net)/.*)(/download)$', path)
  File "/usr/lib64/python2.7/re.py", line 146, in search
    return _compile(pattern, flags).search(string)
```
But this used to work before. `git bisect` says:
```
e3a556cd2a93ec3c45c3627cb16cfcebcfd6b930 is the first bad commit
commit e3a556cd2a93ec3c45c3627cb16cfcebcfd6b930
Author: Massimiliano Culpo <massimiliano.culpo@gmail.com>
Date:   Wed Jul 11 22:12:19 2018 +0200

    Don't propagate version-specific URL overrides to newer versions (#8565)

    When a user specifies a URL for a specific version of a package, Spack originally
    would use that URL for all newer versions of the package. This behavior has
    proven to be generally more harmful than useful, so this PR removes the feature
    such that a version-specific URL override affects only that version.

:040000 040000 7af70e81f82a5d991e3e6dd35190e6dae6d1c4ca df9ab2d340fe459500c308828866514536268745 M      lib
:040000 040000 ffa520bfb42389094d269decc85a9931b2403e41 8a2f431ea0b9e6ed0901c69b9f4aad17c1d19628 M      var
```

Reverting this commit gets the build going
```
balay@cori03:~/spack-tmp> spack install -dirty hypre
==> Error: unrecognized arguments: -dirty
balay@cori03:~/spack-tmp> spack install --dirty hypre
==> intel-mkl@18.0.2.199 : externally installed in /opt/intel
==> intel-mkl@18.0.2.199 : already registered in DB
==> mpich@7.6.0 : has external module in cray-mpich
==> mpich@7.6.0 : is actually installed in /opt/cray/pe/mpt/7.7.0/gni/mpich-intel/16.0
==> mpich@7.6.0 : already registered in DB
==> Installing hypre
```
Also merging https://github.com/adamjstewart/spack/tree/fixes/no-url from #3161 also gets this build going

**Comment by adamjstewart** [comment]
Thanks @balay, I think #3161 will be merged fairly soon so hopefully this problem will go away.
